src/main/java/org/myrobotlab/codec/Api.java

```diff
@@ -142,21 +158,214 @@ static public Message uriToMsg(String uri) {
+  }
 
-    int pos = requestUri.indexOf(Api.PARAMETER_API);
+  static public String getApiKey(String uri) {
+    int pos = uri.indexOf(Api.PARAMETER_API);
     if (pos > -1) {
       pos += Api.PARAMETER_API.length();
-      int pos2 = requestUri.indexOf("/", pos);
+      int pos2 = uri.indexOf("/", pos);
       if (pos2 > -1) {
-        return requestUri.substring(pos, pos2);
+        return uri.substring(pos, pos2);
       } else {
-        return requestUri.substring(pos);
+        return uri.substring(pos);
       }
     }
     return null;
   }
 
-  abstract public Object process(MessageSender sender, OutputStream out, Message msgFromUri, String data) throws Exception;
+  // Api Imp has opportunity to override
+  protected void handleSession(WebGui webgui, String apiKey, AtmosphereResource r) {
+    AtmosphereRequest request = r.getRequest();
+    HttpSession s = request.getSession(true);
+    if (s != null) {
+      // System.out.println("put session uuid {}", r.uuid());
+      webgui.getSessions().put(r.uuid(), request.getSession(true));
+    }
+  }
+
+  protected void setBroadcaster(WebGui webgui, String apiKey, AtmosphereResource r) {
+    // FIXME - maintain single broadcaster for each session ?    
+    r.setBroadcaster(webgui.getBroadcaster());
+  }
+
+ 
+  protected void setHeaderContentType(WebGui webgui, String apiKey, AtmosphereResource r) {
+    r.getResponse().addHeader("Content-Type", CodecUtils.MIME_TYPE_JSON);
+  }
+
+  protected void suspend(WebGui webgui, String apiKey, AtmosphereResource r) {
+    if (!apiKey.equals(ApiFactory.API_TYPE_SERVICE) && !r.isSuspended()) {
+      r.suspend();
+    }
+  }
+
+  protected void handleError(WebGui webgui, String apiKey, AtmosphereResource r, Exception e) {
+    
+    try {
+
+      AtmosphereResponse response = r.getResponse();
+      OutputStream out = response.getOutputStream();
+
+      response.addHeader("Content-Type", CodecUtils.MIME_TYPE_JSON);
+      Codec codec = CodecFactory.getCodec(CodecUtils.MIME_TYPE_JSON);
+
+      Status error = Status.error(e);   // name == null anonymous ?
+      Message msg = Message.createMessage(webgui, null, CodecUtils.getCallbackTopicName("getStatus"), error);
+      if (ApiFactory.API_TYPE_SERVICE.equals(apiKey)) {
+        // for the purpose of only returning the data
+        // e.g. http://api/services/runtime/getUptime -> return the uptime
+        // only not the message
+        if (msg.data == null) {
+          log.warn("<< {}", CodecJson.encode(null));
+          codec.encode(out, null);
+        } else {
+          // return the return type
+          log.warn("<< {}", CodecJson.encode(msg.data[0]));
+          codec.encode(out, msg.data[0]);
+        }
+      } else {
+        // API_TYPE_MESSAGES
+        // DEPRECATE - FOR LOGGING ONLY REMOVE
+        log.warn("<< {}", CodecJson.encode(msg)); // FIXME if logTraffic
+        codec.encode(out, msg);
+      }
+
+    } catch (Exception e2) {
+      log.error("respond threw", e);

```

Comment:
we should be logging "e2" here not "e"

Reply:
Maybe, log the exception of trying to report the exception ?   Sounds pretty chatty, but okay, added e2 locally.